File: packages/coding-agent/src/tools/computer/prelude.py
def _make_computer():
    import re
    from types import MappingProxyType

    def _encode_arg(value):
        if isinstance(value, MappingProxyType):
            return dict(value)
        if isinstance(value, re.Pattern):
            if not isinstance(value.pattern, str):
                raise TypeError("computer helpers require regular expressions with string patterns")
            flags = ""
            if value.flags & re.IGNORECASE:
                flags += "i"
            if value.flags & re.MULTILINE:
                flags += "m"
            if value.flags & re.DOTALL:
                flags += "s"
            return {"__omp_re": {"source": value.pattern, "flags": flags}}
        return value

    def _arguments(args, kwargs):
        values = list(args)
        while values and values[-1] is None:
            values.pop()
        values = [_encode_arg(value) for value in values]
        options = {
            key: _encode_arg(value)
            for key, value in kwargs.items()
            if value is not None
        }
        if options:
            values.append(options)
        return values

    async def _invoke(action, options):
        response = await _omp_prelude(
            "computer",
            {
                **{
                    key: value
                    for key, value in options.items()
                    if value is not None
                },
                "action": action,
            },
        )
        if isinstance(response, str):
            return {}
        if not isinstance(response, dict):
            raise RuntimeError("computer returned an invalid response")
        text = response.get("text")
        if isinstance(text, str) and text:
            print(text)
        details = response.get("details")
        return details if isinstance(details, dict) else {}

    async def _call(chain):
        details = await _invoke("call", {"chain": chain})
        return details.get("value")

    def _step(method, args, kwargs):
        return {"method": method, "args": _arguments(args, kwargs)}

    class _Element:
        __slots__ = ("ref", "role", "subrole", "label", "value", "placeholder", "help", "description", "enabled", "selected", "actions", "bounds", "pid", "windowId", "_owner")

        def __init__(self, snapshot, identity=None):
            for field in _Element.__slots__:
                if field == "_owner":
                    continue
                value = snapshot.get(field)
                if field == "bounds" and isinstance(value, dict):
                    value = MappingProxyType(dict(value))
                if field == "actions" and isinstance(value, list):
                    value = tuple(value)
                object.__setattr__(self, field, value)
            owner = identity
            if owner is None and self.windowId is not None and self.pid is not None:
                owner = {"id": self.windowId, "pid": self.pid}
            object.__setattr__(self, "_owner", MappingProxyType(dict(owner)) if owner is not None else None)

        def __setattr__(self, name, value):
            raise AttributeError("computer element fields are immutable observation data")

        def __repr__(self):
            return f"<computer.Element ref={self.ref!r} role={self.role!r}>"

        async def _method(self, method, args, kwargs):
            chain = [_step("window", (dict(self._owner),), {})] if self._owner is not None else []
            return await _call([*chain, _step("ref", (self.ref,), {}), _step(method, args, kwargs)])

        async def setValue(self, *args, **kwargs):
            return await self._method("setValue", args, kwargs)

        async def perform(self, *args, **kwargs):
            return await self._method("perform", args, kwargs)

        async def press(self, *args, **kwargs):
            return await self._method("press", args, kwargs)

        async def click(self, *args, **kwargs):
            return await self._method("click", args, kwargs)

        async def doubleClick(self, *args, **kwargs):
            return await self._method("doubleClick", args, kwargs)

        async def type(self, *args, **kwargs):
            return await self._method("type", args, kwargs)

        async def scroll(self, *args, **kwargs):
            return await self._method("scroll", args, kwargs)


    class _LazyElement(_Element):
        """An `_Element` built from a bare ref; awaiting it fetches the full snapshot."""
        __slots__ = ("_resolve",)

        def __init__(self, snapshot, identity, resolve):
            super().__init__(snapshot, identity)
            object.__setattr__(self, "_resolve", resolve)

        def __await__(self):
            return self._resolve().__await__()

    class _Window:
        __slots__ = ("id", "app", "title", "pid", "bounds", "onScreen", "layer", "zIndex", "kind", "initialObservation", "inspectionError", "initialScreenshot", "screenshotError")

        def __init__(self, snapshot):
            if not isinstance(snapshot.get("id"), str) or type(snapshot.get("pid")) is not int:
                raise TypeError("computer window snapshot requires an exact id and PID")
            for field in self.__slots__:
                value = snapshot.get(field)
                if field == "bounds" and isinstance(value, dict):
                    value = MappingProxyType(dict(value))
                object.__setattr__(self, field, value)

        def __setattr__(self, name, value):
            raise AttributeError("computer window fields are immutable observation data")

        def __repr__(self):
            return f"<computer.Window id={self.id!r} app={self.app!r}>"

        async def _method(self, method, args, kwargs):
            return await _call([_step("window", ({"id": self.id, "pid": self.pid},), {}), _step(method, args, kwargs)])

        async def screenshot(self, *args, **kwargs):
            return await self._method("screenshot", args, kwargs)

        async def click(self, *args, **kwargs):
            return await self._method("click", args, kwargs)

        async def doubleClick(self, *args, **kwargs):
            return await self._method("doubleClick", args, kwargs)

        async def hover(self, *args, **kwargs):
            return await self._method("hover", args, kwargs)

        async def drag(self, *args, **kwargs):
            return await self._method("drag", args, kwargs)

        async def scroll(self, *args, **kwargs):
            return await self._method("scroll", args, kwargs)

        async def type(self, *args, **kwargs):
            return await self._method("type", args, kwargs)

        async def press(self, *args, **kwargs):
            return await self._method("press", args, kwargs)

        async def reveal(self, *args, **kwargs):
            return await self._method("reveal", args, kwargs)

        async def observe(self, *args, **kwargs):
            return await self._method("observe", args, kwargs)

        async def setValue(self, *args, **kwargs):
            return await self._method("setValue", args, kwargs)

        async def setFrame(self, *args, **kwargs):
            return await self._method("setFrame", args, kwargs)

        async def menu(self, *args, **kwargs):
            return await self._method("menu", args, kwargs)

        async def verify(self, *args, **kwargs):
            return await _call([_step("verifyWindow", ({"id": self.id, "pid": self.pid}, *args), kwargs)])

        async def find(self, *args, **kwargs):
            return [_Element(snapshot, {"id": self.id, "pid": self.pid}) for snapshot in await self._method("find", args, kwargs)]

        def ref(self, ref):
            """A ref handle: act directly (`await win.ref(r).click()`) or `await win.ref(r)` for the snapshot."""
            identity = {"id": self.id, "pid": self.pid}

            async def resolve():
                snapshot = await self._method("ref", (ref,), {})
                return _Element(snapshot, identity) if isinstance(snapshot, dict) else None

            return _LazyElement({"ref": ref}, identity, resolve)

    class _Clipboard:
        __slots__ = ()

        async def read(self):
            return await _call([_step("clipboard.read", (), {})])

        async def write(self, text):
            return await _call([_step("clipboard.write", (text,), {})])

    class _Computer:
        __slots__ = ("clipboard",)

        def __init__(self):
            self.clipboard = _Clipboard()

        def __repr__(self):
            return "<computer>"

        async def _method(self, method, args, kwargs):
            return await _call([_step(method, args, kwargs)])

        async def apps(self, *args, **kwargs):
            return await self._method("apps", args, kwargs)

        async def launch(self, *args, **kwargs):
            return await self._method("launch", args, kwargs)

        async def displays(self, *args, **kwargs):
            return await self._method("displays", args, kwargs)

        async def windows(self, *args, **kwargs):
            return await self._method("windows", args, kwargs)

        async def screenshot(self, *args, **kwargs):
            return await self._method("screenshot", args, kwargs)

        async def click(self, *args, **kwargs):
            return await self._method("click", args, kwargs)

        async def doubleClick(self, *args, **kwargs):
            return await self._method("doubleClick", args, kwargs)

        async def move(self, *args, **kwargs):
            return await self._method("move", args, kwargs)

        async def drag(self, *args, **kwargs):
            return await self._method("drag", args, kwargs)

        async def scroll(self, *args, **kwargs):
            return await self._method("scroll", args, kwargs)

        async def type(self, *args, **kwargs):
            return await self._method("type", args, kwargs)

        async def press(self, *args, **kwargs):
            return await self._method("press", args, kwargs)

        async def window(self, *args, launch=None, ambiguous=None, screenshot=None, silent=None, maxDepth=None, maxElements=None, query=None, **kwargs):
            """Acquire one exact window and its initial background inspection."""
            selectors = _arguments(args, kwargs)
            if len(selectors) != 1:
                raise TypeError("computer.window expects one selector or filter keywords")
            options = {key: value for key, value in {"launch": launch, "ambiguous": ambiguous, "screenshot": screenshot, "silent": silent, "maxDepth": maxDepth, "maxElements": maxElements, "query": query}.items() if value is not None}
            snapshot = await self._method("acquireWindow", (selectors[0], options), {})
            return _Window(snapshot) if isinstance(snapshot, dict) else None

        async def focusedWindow(self):
            snapshot = await self._method("focusedWindow", (), {})
            return _Window(snapshot) if isinstance(snapshot, dict) else None

        def ref(self, ref):
            """A ref handle: act directly or await it for snapshot data (not native liveness)."""

            async def resolve():
                snapshot = await self._method("ref", (ref,), {})
                return _Element(snapshot) if isinstance(snapshot, dict) else None

            return _LazyElement({"ref": ref}, None, resolve)

        async def run(self, code, *, read_only=None, timeout=None):
            """Run a JavaScript code string in the persistent desktop session and return its value."""
            if not isinstance(code, str):
                raise TypeError("computer.run() expects a JavaScript code string")
            details = await _invoke(
                "run",
                {"code": code, "read_only": read_only, "timeout": timeout},
            )
            return details.get("value")

        async def capabilities(self):
            """Initialize the desktop and return backend capabilities and permission state."""
            return await self._method("capabilities", (), {})

        async def help(self):
            """Print the typed computer API: every interface and signature the handles expose."""
            await _invoke("help", {})

        async def release(self):
            """Drain and release capture/control resources; later calls start a fresh desktop worker."""
            await _invoke("release", {})

        async def close(self):
            """End the persistent desktop session; later calls fail."""
            await _invoke("close", {})

    return _Computer()


computer = _make_computer()
del _make_computer

File: packages/coding-agent/src/tools/computer/prelude.js
{
	const validateOptions = (label, options) => {
		if (options === undefined) return {};
		if (options === null || typeof options !== "object" || Array.isArray(options)) {
			throw new TypeError(`${label}() expects an options object`);
		}
		return options;
	};
	const serializeFunction = (label, fn) => {
		const source = String(fn);
		if (source.includes("[native code]")) {
			throw new TypeError(
				`${label} cannot serialize a native or bound function; pass an arrow or function expression`,
			);
		}
		return source;
	};
	const encodeArg = (label, value) => {
		if (typeof value === "function") return { __omp_fn: serializeFunction(label, value) };
		if (value instanceof RegExp) return { __omp_re: { source: value.source, flags: value.flags } };
		return value;
	};
	const encodeArgs = (label, args) => {
		const trimmed = [...args];
		while (trimmed.length > 0 && trimmed[trimmed.length - 1] === undefined) trimmed.pop();
		return trimmed.map(value => encodeArg(label, value));
	};
	const invoke = async (action, options) => {
		const response = await globalThis.__omp_prelude__("computer", { ...options, action });
		if (response && typeof response.text === "string" && response.text.length > 0) {
			globalThis.__omp_display__(response.text);
		}
		const details =
			response && typeof response.details === "object" && response.details !== null ? response.details : {};
		// The host already rendered this value (an observation tree) into the
		// text above; the cell must not print it a second time as its trailing
		// expression.
		if (details.rendered === true) globalThis.__omp_presented__?.(details.value);
		return details;
	};
	const callValue = async chain => {
		const details = await invoke("call", { chain });
		return details.value;
	};
	const step = (method, args) => ({ method, args: encodeArgs("computer helper argument", args) });
	// Methods stay non-enumerable so handles display and serialize as their identity fields only.
	const defineMethod = (target, name, fn) => Object.defineProperty(target, name, { value: fn });
	const defineValueMethods = (target, methods, chain) => {
		for (const method of methods) {
			defineMethod(target, method, (...args) => callValue(chain(step(method, args))));
		}
	};

	const windowFields = ["id", "app", "title", "pid", "bounds", "onScreen", "layer", "zIndex", "kind"];
	const windowValueMethods = [
		"screenshot",
		"click",
		"doubleClick",
		"hover",
		"drag",
		"scroll",
		"type",
		"press",
		"reveal",
		"observe",
		"setValue",
		"setFrame",
		"menu",
	];
	const elementFields = [
		"ref",
		"role",
		"subrole",
		"label",
		"value",
		"placeholder",
		"help",
		"description",
		"enabled",
		"selected",
		"actions",
		"bounds",
		"pid",
		"windowId",
	];
	const elementValueMethods = ["click", "doubleClick", "setValue", "type", "press", "scroll", "perform"];
	const desktopValueMethods = [
		"capabilities",
		"apps",
		"launch",
		"displays",
		"windows",
		"screenshot",
		"click",
		"doubleClick",
		"move",
		"drag",
		"scroll",
		"type",
		"press",
	];

	const copyFields = (target, fields, snapshot) => {
		for (const field of fields) {
			if (snapshot[field] !== undefined) {
				let value = snapshot[field];
				if (field === "bounds" && value !== null) value = Object.freeze({ ...value });
				if (field === "actions" && Array.isArray(value)) value = Object.freeze([...value]);
				Object.defineProperty(target, field, { value, enumerable: true });
			}
		}
	};
	const makeElement = (snapshot, identity) => {
		const element = {};
		copyFields(element, elementFields, snapshot);
		defineMethod(element, "toString", () => `<element ${snapshot.ref} ${snapshot.role}>`);
		const owner =
			identity ??
			(snapshot.windowId !== undefined && snapshot.pid !== undefined
				? { id: snapshot.windowId, pid: snapshot.pid }
				: undefined);
		const via = next => [...(owner ? [step("window", [owner])] : []), step("ref", [snapshot.ref]), next];
		defineValueMethods(element, elementValueMethods, via);
		return Object.freeze(element);
	};
	const resolveElement = async (chain, identity) => {
		const snapshot = await callValue(chain);
		return snapshot ? makeElement(snapshot, identity) : null;
	};
	/**
	 * A ref handle that acts without a round trip (`win.ref(r).click()`) and
	 * still resolves to the full snapshot when awaited (`await win.ref(r)`).
	 */
	const lazyElement = (chain, ref, identity) => {
		const element = makeElement({ ref }, identity);
		const resolved = () => resolveElement(chain, identity);
		return Object.freeze(
			Object.create(element, {
				then: { value: (onFulfilled, onRejected) => resolved().then(onFulfilled, onRejected) },
			}),
		);
	};
	const makeWindow = snapshot => {
		if (typeof snapshot.id !== "string" || !Number.isInteger(snapshot.pid)) {
			throw new TypeError("computer window snapshot requires an exact id and PID");
		}
		const win = {};
		copyFields(win, windowFields, snapshot);
		for (const field of ["initialObservation", "inspectionError", "initialScreenshot", "screenshotError"]) {
			if (snapshot[field] !== undefined) Object.defineProperty(win, field, { value: snapshot[field] });
		}
		defineMethod(win, "toString", () => `<window ${snapshot.id} ${snapshot.app}>`);
		const identity = { id: snapshot.id, pid: snapshot.pid };
		const via = next => [step("window", [identity]), next];
		defineValueMethods(win, windowValueMethods, via);
		defineMethod(win, "verify", (...args) => callValue([step("verifyWindow", [identity, ...args])]));
		defineMethod(win, "find", async query =>
			(await callValue(via(step("find", [query])))).map(item => makeElement(item, identity)),
		);
		defineMethod(win, "ref", ref => lazyElement(via(step("ref", [ref])), ref, identity));
		// @achieve
		// The chooser sub-loop runs on the host, where the judge and the typed
		// observation live; the handle only names the goal and the bounds.
		defineMethod(win, "achieve", async (goal, options) => {
			if (typeof goal !== "string" || goal.length === 0) {
				throw new TypeError("win.achieve() expects a non-empty goal string");
			}
			const opts = validateOptions("win.achieve", options);
			const parameters = { window: identity, goal };
			if (opts.maxSteps !== undefined) parameters.maxSteps = opts.maxSteps;
			if (opts.confidence !== undefined) parameters.confidence = opts.confidence;
			const details = await invoke("achieve", parameters);
			return details.value;
		});
		// @end achieve
		return Object.freeze(win);
	};
	const resolveWindow = async chain => {
		const snapshot = await callValue(chain);
		return snapshot ? makeWindow(snapshot) : null;
	};

	const computer = {};
	defineValueMethods(computer, desktopValueMethods, next => [next]);
	computer.window = (selector, options) =>
		resolveWindow([step("acquireWindow", [selector, validateOptions("computer.window", options)])]);
	computer.focusedWindow = () => resolveWindow([step("focusedWindow", [])]);
	computer.ref = ref => lazyElement([step("ref", [ref])], ref);
	computer.clipboard = Object.freeze({
		read: () => callValue([step("clipboard.read", [])]),
		write: text => callValue([step("clipboard.write", [text])]),
	});
	computer.run = async (fnOrCode, options) => {
		if (typeof fnOrCode !== "function" && typeof fnOrCode !== "string") {
			throw new TypeError("computer.run() expects a function or code string");
		}
		const opts = validateOptions("computer.run", options);
		const parameters = {};
		if (opts.read_only !== undefined) parameters.read_only = opts.read_only;
		if (opts.timeout !== undefined) parameters.timeout = opts.timeout;
		if (typeof fnOrCode === "function") {
			parameters.fn = serializeFunction("computer.run()", fnOrCode);
			parameters.args = encodeArgs("computer.run() argument", Array.isArray(opts.args) ? opts.args : []);
		} else {
			parameters.code = fnOrCode;
		}
		const details = await invoke("run", parameters);
		return details.value;
	};
	// The host prints the typed API; returning it too would print it twice.
	computer.help = async () => {
		await invoke("help", {});
	};
	computer.release = async () => {
		await invoke("release", {});
	};
	computer.close = async () => {
		await invoke("close", {});
	};
	globalThis.computer = Object.freeze(computer);
}
